cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// The translation unit for reduction `minmax`
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform_reduce.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief stores the minimum and maximum
+ * values that have been encountered so far
+ *
+ */
+template <typename T>
+struct minmax_pair {
+  T min_val;
+  T max_val;
+
+  __host__ __device__ minmax_pair()
+    : min_val(cudf::DeviceMin::identity<T>()), max_val(cudf::DeviceMax::identity<T>()){};
+  __host__ __device__ minmax_pair(T val) : min_val(val), max_val(val){};
+  __host__ __device__ minmax_pair(T min_val_, T max_val_) : min_val(min_val_), max_val(max_val_){};
+};
+
+/**
+ * @brief Reduce the binary operation in device and return a device scalar.
+ *
+ * @tparam Op Binary operator functor
+ * @tparam InputIterator Input iterator Type
+ * @param d_in input iterator
+ * @param num_items number of items to reduce
+ * @param binary_op binary operator used to reduce
+ * @param mr Device resource used for result allocation
+ * @param stream CUDA stream to run kernels on.
+ * @return rmm::device_scalar<OutputType>
+ */
+template <typename T, typename Op, typename InputIterator, typename OutputType = minmax_pair<T>>
+// typename thrust::iterator_value<InputIterator>::type>
+// above produces compilation error where OutputType is somehow int
+auto reduce_device(InputIterator d_in,
+                   cudf::size_type num_items,
+                   Op binary_op,
+                   rmm::mr::device_memory_resource *mr,
+                   cudaStream_t stream)  //-> typename std::enable_if_t<std::is_same<T, int>::value, rmm::device_scalar<OutputType>>
+{
+  OutputType identity{};
```

Comment:
One issue may be that `identity` declared here has value 0. The identity value is included in DeviceReduce operation. And if all you values are `x >= 0`  then `min(x,0)=0` -- always returns the identity. But I'm not sure why the max-half of the `minmax` would return identity too.

Reply:
Thank you for looking at this, but wouldn't this be `minmax_pair<T>` so it would be a pair with min value of `cudf::DeviceMin::identity<T>()` and max value of `cudf::DeviceMax::identity<T>()`? I think this is ok here. The return value is actually min value of max for type and max value of min for type. So it seems to be returning the `identity` variable here.

Potentially interesting and a hint is if I call:
```
  OutputType tresult = thrust::reduce(rmm::exec_policy(stream)->on(stream), d_in, d_in + num_items, identity, [] __device__ (OutputType const &lhs, OutputType const &rhs ) {
    return OutputType{thrust::min(lhs.min_val, rhs.min_val), thrust::max(lhs.max_val, rhs.max_val)};
```
I get this error:

```
unknown file: Failure
C++ exception with description "after reduction step 2: cudaErrorInvalidDeviceFunction: invalid device function" thrown in the test body.
```

Reply:
Not sure if this is chasing a red herring or not, but:
```
auto works = thrust::transform_reduce(rmm::exec_policy(stream)->on(stream), d_in, d_in + num_items, [] __device__ (auto e) { return e; }, minmax_pair<T>{}, binary_op);
auto crashes = thrust::reduce(rmm::exec_policy(stream)->on(stream), d_in, d_in + num_items, minmax_pair<T>{}, binary_op);
```